### PR TITLE
Bug 1937972: router/template: Cache compiled regular expressions

### PR DESCRIPTION
**Pull request by frobware**
This is an optimisation that caches compiled regular expressions.

When the template is written calls to `matchString` and `firstMatch`
would recompile the regular expression on every call. When you are
processing ~13,000 routes the time to compile on each invocation becomes
significant. This change brings a reduction of 60% when processing and
writing `haproxy.config`.

```console
router.go:456] template "msg"="writeConfig"  "duration"="10.663160501s"
router.go:456] template "msg"="writeConfig"  "duration"="10.328840858s"
router.go:456] template "msg"="writeConfig"  "duration"="10.762876672s"
```

And if we cache the REs:

```console
template "msg"="writeConfig"  "duration"="3.659229627s"
template "msg"="writeConfig"  "duration"="3.718457818s"
template "msg"="writeConfig"  "duration"="3.672131941s"
```

Related to: https://bugzilla.redhat.com/show_bug.cgi?id=1929821

### CODE REVIEW

**Review by knobunc — APPROVED** [review]

**Comment by knobunc** [comment]
Looks good to me!  But waiting for a second reviewer.

File: pkg/router/template/template_helper.go
Comment (Miciah) [inline pkg/router/template/template_helper.go:108]: `matchPattern` isn't used in the template, so this shouldn't affect performance (but it shouldn't hurt anything either).

File: pkg/router/template/template_helper.go
Comment (Miciah) [inline pkg/router/template/template_helper.go:35]: I was wondering if we needed garbage collection for this map.  I believe we do not as every use of `firstMatch` specifies a pattern that is a hard-coded in the template itself (as opposed to specifying a pattern from an annotation, for example), so the size of the map is bounded by a function of the static set of patterns that are used in `haproxy-config.template`.

File: pkg/router/template/template_helper.go
Comment (Miciah) [inline pkg/router/template/template_helper.go:37]: Would you mind adding some godoc?

**Review by Miciah — COMMENTED** [review]
This is terrific.  
/lgtm
/hold
in case you want to wait for further reviews or do any follow-up work such as comments or adding a BZ reference.

**Comment by sgreene570** [comment]
/retitle Bug 1937972: router/template: Cache compiled regular expressions

File: pkg/router/template/template_helper.go
Comment (frobware) [reply pkg/router/template/template_helper.go:108]: This was explicit as `MatchString` compiles a regexp too:

```go
// MatchString reports whether the string s
// contains any match of the regular expression pattern.
// More complicated queries need to use Compile and the full Regexp interface.
func MatchString(pattern string, s string) (matched bool, err error) {
	re, err := Compile(pattern)
	if err != nil {
		return false, err
	}
	return re.MatchString(s), nil
}
```

File: pkg/router/template/template_helper.go
Comment (frobware) [reply pkg/router/template/template_helper.go:47]: Force-pushed.

File: pkg/router/template/template_helper.go
Comment (Miciah) [reply pkg/router/template/template_helper.go:120]: I mean, this entire function is effectively dead code anyway, unless someone is using a custom template that uses `matchPattern`.  That said, it is possible that someone _is_ using a custom template that uses `matchPattern`, so we might as well optimize it.

File: pkg/router/template/template_helper.go
Comment (Miciah) [inline pkg/router/template/template_helper.go:62]: What's this line for?  Copy and paste error?

Comment (frobware) [reply pkg/router/template/template_helper.go:62]: Errr. yes. Removing. :)

**Pull request merged**
